vite config files can't be compiled without `dom` types

### Describe the bug

when attempting to compile a `vite.config.ts` file without dom types, and without `skipLibCheck` in `tsconfig.json`, the following errors occur:

```
❯ tsc
node_modules/esbuild/lib/main.d.ts:592:16 - error TS2503: Cannot find namespace 'WebAssembly'.

592   wasmModule?: WebAssembly.Module
                   ~~~~~~~~~~~

node_modules/vite/dist/node/index.d.ts:1259:13 - error TS2304: Cannot find name 'Worker'.

1259     worker: Worker
                 ~~~~~~


Found 2 errors in 2 files.

Errors  Files
     1  node_modules/esbuild/lib/main.d.ts:592
     1  node_modules/vite/dist/node/index.d.ts:1259
```

i believe it's because these types are defined globally in `lib.dom.d.ts`. but since the `vite.config.ts` file is executed by nodejs, it's incorrect to include the dom types when compiling it.

is it possible to resolve this by importing the `Worker` type from elsewhere?

- esbuild issue: https://github.com/evanw/esbuild/issues/2388
- typescript issue: https://github.com/microsoft/TypeScript/issues/50424

### Reproduction

https://stackblitz.com/edit/vitejs-vite-ghpgs3?file=vite.config.ts,tsconfig.json,package.json&terminal=dev

### System Info

```shell
❯ npx envinfo --system --npmPackages '{vite,@vitejs/*}' --binaries --browsers
ERR!: Failed to install package

error Command failed with exit code 1
```


### Used Package Manager

npm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/tsconfig.check.json
+++ b/packages/vite/tsconfig.check.json
@@ -3,7 +3,7 @@
     "target": "ES2020",
     "moduleResolution": "node16",
     "module": "Node16",
-    "lib": ["ES2020", "WebWorker"], // ES2020 is very conservative check for client types, could be bumped if needed
+    "lib": ["ES2020"], // ES2020 is very conservative check for client types, could be bumped if needed
     "types": [], // Avoid checking unrelated node_modules types
     "noEmit": true,
     "strict": true,

--- a/packages/vite/types/importGlob.d.ts
+++ b/packages/vite/types/importGlob.d.ts
@@ -42,6 +42,7 @@ export type GeneralImportGlobOptions = ImportGlobOptions<boolean, string>
  * the Worker interface will be merged correctly.
  */
 declare global {
+  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
   interface Worker {}
 }
 

--- a/packages/vite/types/importGlob.d.ts
+++ b/packages/vite/types/importGlob.d.ts
@@ -36,7 +36,11 @@ export interface ImportGlobOptions<
 
 export type GeneralImportGlobOptions = ImportGlobOptions<boolean, string>
 
-// provide a fallback in case the Worker interface isn't provided in the environment
+/**
+ * Declare Worker in case DOM is not added to the tsconfig lib causing
+ * Worker interface is not defined. For developers with DOM lib added,
+ * the Worker interface will be merged correctly.
+ */
 declare global {
   interface Worker {}
 }

--- a/packages/vite/types/importGlob.d.ts
+++ b/packages/vite/types/importGlob.d.ts
@@ -36,6 +36,11 @@ export interface ImportGlobOptions<
 
 export type GeneralImportGlobOptions = ImportGlobOptions<boolean, string>
 
+// provide a fallback in case the Worker interface isn't provided in the environment
+declare global {
+  interface Worker {}
+}
+
 export interface KnownAsTypeMap {
   raw: string
   url: string

